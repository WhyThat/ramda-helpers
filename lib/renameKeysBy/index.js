import {
  adjust,
  curry,
  fromPairs,
  identity,
  ifElse,
  map,
  pipe,
  toPairs,
  when,
} from 'ramda';
import isArray from '../isArray';
import isPlainObject from '../isPlainObject';

// eslint-disable-next-line no-use-before-define
const transformObject = strFormatter => ifElse(isPlainObject, renameKeysBy(strFormatter), identity);
const transformArray = strFormatter => map(transformObject(strFormatter));

const transformKey = strFormatter => adjust(strFormatter, 0);
const transformValue = strFormatter => adjust(
  pipe(
    when(isArray, transformArray(strFormatter)),
    transformObject(strFormatter)), 1);

const renameKeysBy = curry(
  (strFormatter, obj) => pipe(
    toPairs,
    map(pipe(transformKey(strFormatter), transformValue(strFormatter))),
    fromPairs)(obj));

/**
 * Returns objects with all keys renamed by the predicate (recursivly)
 * @name renameKeysBy
 * @param {function} predicate
 * @param {Object} obj
 * @returns {Object}
 * @example
 * renameByKey(str => `prefix_${str}`, { test: 'test' })
 *    //=> { prefix_test: 'test' }
 */
export default renameKeysBy;
