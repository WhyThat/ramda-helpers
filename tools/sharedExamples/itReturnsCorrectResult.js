export default (result) => {
  test('returns correct result', () => {
    expect(result).toMatchSnapshot();
  });
};
