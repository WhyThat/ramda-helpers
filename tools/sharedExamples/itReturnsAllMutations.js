import itReturnsCorrectResult from './itReturnsCorrectResult';

export default (fn, mutations) => {
  mutations.forEach((mutation, index) => {
    const name = mutation.name || `mutations ${index}`;

    describe(name, () => {
      itReturnsCorrectResult(fn(mutation.props));
    });
  });
};
