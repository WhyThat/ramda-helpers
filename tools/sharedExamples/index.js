
import itReturnsAllMutations from './itReturnsAllMutations';
import itReturnsCorrectResult from './itReturnsCorrectResult';

export {
  itReturnsAllMutations,
  itReturnsCorrectResult,
};
