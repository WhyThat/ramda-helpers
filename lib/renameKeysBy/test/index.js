import { itReturnsAllMutations } from '../../test/sharedExamples';
import renameKeysBy from '../';

const renameWithPrefix = renameKeysBy(str => `prefix_${str}`);
describe('renameKeysBy', () => {
  const mutations = [{
    name: 'With empty object',
    props: {},
  }, {
    name: 'With simple object',
    props: {
      test: '',
    },
  }, {
    name: 'With nested object',
    props: {
      test: {
        test2: '',
      },
    },
  }, {
    name: 'With complex object',
    props: {
      test: {
        test2: [
          { test2_1: [{ test2_1_1: '' }] },
          { test2_2: '' },
        ],
      },
    },
  }];
  itReturnsAllMutations(renameWithPrefix, mutations);
});
