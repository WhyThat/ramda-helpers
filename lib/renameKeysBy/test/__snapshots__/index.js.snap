exports[`renameKeysBy With complex object returns correct result 1`] = `
Object {
  "prefix_test": Object {
    "prefix_test2": Array [
      Object {
        "prefix_test2_1": Array [
          Object {
            "prefix_test2_1_1": "",
          },
        ],
      },
      Object {
        "prefix_test2_2": "",
      },
    ],
  },
}
`;

exports[`renameKeysBy With empty object returns correct result 1`] = `Object {}`;

exports[`renameKeysBy With nested object returns correct result 1`] = `
Object {
  "prefix_test": Object {
    "prefix_test2": "",
  },
}
`;

exports[`renameKeysBy With simple object returns correct result 1`] = `
Object {
  "prefix_test": "",
}
`;
